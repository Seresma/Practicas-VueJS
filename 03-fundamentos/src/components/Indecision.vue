<!-- CTRL + k + c para comentar -->
<!-- CTRL + k + u para descomentar -->

<template>
  <!-- Para v-bind shortcut : se utiliza para bindear atributos -->
    <!-- v-if para que solo se muestre en caso de que no sea undefined -->
  <img v-if="img" :src="img" alt="bg">
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input v-model="question" type="text" placeholder="Hazme una pregunta">
    <p>Recuerda terminar con un signo de interrogacion (?)</p>
    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ answer === "yes" ? "SÃ­"  : "No" }}</h1>
    </div>
  </div>
</template>

<script>
export default {

  // Objetos reactivos
  // Info del componente
  data() {
    return {
      question: null,
      answer: null,
      img: null,
      isValidQuestion: false
    }
  },

  methods: {
    async getAnswer() {
      this.answer = "Pensando..."

      const {answer, image} = await fetch("https://yesno.wtf/api").then(r => r.json())

      this.answer = answer
      this.img = image
    },
  },

  // Para observarbles en ciertos objetos se tiene que llamar igual que la propiedad
  watch: {
    question(value, oldValue) {
      
      this.isValidQuestion = false

      if (!value.includes('?')) return

      this.isValidQuestion = true

      this.getAnswer()

    }
  }
}
</script>

<!-- scoped para que sea estilo del propio componente y no de los estilos de App.vue -->
<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}

input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>