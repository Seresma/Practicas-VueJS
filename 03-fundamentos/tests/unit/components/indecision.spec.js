import { shallowMount } from "@vue/test-utils"
import Indecision from "@/components/Indecision"

describe("Indecision Component", () => {

    let wrapper
    let clgSpy
    let getAnswerSpy

    // Mockeamos fetch ya que solo existe en el navegador
    // Ejemplo de respuesta de la API
    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: "yes",
            forced: false,
            image: "https://yesno.wtf/assets/yes/2.gif"
        })
    }))

    beforeEach(() => {
        wrapper = shallowMount(Indecision)
        clgSpy = jest.spyOn(console, "log")
        // vm es la instancia de vue
        getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer")

        jest.clearAllMocks()
    })

    // SIEMPRE QUE SE HAGAN MODIFICACIONES REACTIVAS (DOM) PONER ASYNC/AWAIT
    test("debe de hacer match con el snapshot", () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test("escribir en el input no debe de dispara nada (console.log)", async () => {

        const input = wrapper.find("input")
        await input.setValue("Hola mundo")

        expect(clgSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).not.toHaveBeenCalled()

    })

    test("escribir el simbolo '?' debe de dispara el getAnswer", async () => {

        const input = wrapper.find("input")
        await input.setValue("Hola mundo?")

        expect(clgSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).toHaveBeenCalledTimes(1)
    })

    test("pruebas en getAnswer", async () => {
        await wrapper.vm.getAnswer()

        const img = wrapper.find("img")

        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif")
        expect(wrapper.vm.answer).toBe("SÃ­")

    })

    test("pruebas en getAnswer - fallo en la API", async () => {

        // Simulamos el fallo de la API
        fetch.mockImplementationOnce(() => Promise.reject("API is down"))
        
        await wrapper.vm.getAnswer()

        const img = wrapper.find("img")

        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe("No se pudo cargar de la API")
    })




})