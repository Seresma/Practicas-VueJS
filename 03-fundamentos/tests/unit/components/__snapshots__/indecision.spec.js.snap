// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component debe de hacer match con el snapshot 1`] = `
<!-- Para v-bind shortcut : se utiliza para bindear atributos -->
<!-- v-if para que solo se muestre en caso de que no sea undefined -->
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con un signo de interrogacion (?)</p>
  <!--v-if-->
</div>
`;
